openlibrary/core/community_batch_imports.py

```diff
@@ -0,0 +1,140 @@
+
+from json.decoder import JSONDecodeError
+
+from pydantic import ValidationError
+from pydantic_core import ErrorDetails
+from openlibrary import accounts
+from openlibrary.core.imports import Batch
+from openlibrary.plugins.importapi.import_edition_builder import import_edition_builder
+from openlibrary.catalog.add_book import (
+    IndependentlyPublished,
+    PublicationYearTooOld,
+    PublishedInFutureYear,
+    SourceNeedsISBN,
+    validate_record,
+)
+
+
+def generate_hash(data: bytes, length: int = 20):
+    """
+    Generate a SHA256 hash of data and truncate it to length.
+
+    This is used to help uniquely identify a community_batch_import. Truncating
+    to 20 characters gives about a 50% chance of collision after 1 billion
+    community hash imports. According to ChatGPT, anyway. :)
+    """
+    hash = hashlib.sha256()
+    hash.update(data)
+    return hash.hexdigest()[:length]
+
+
+@dataclass
+class CbiErrorItem:
+    """
+    Represents a Community Batch Import error item, containing a line number and error message.
+
+    As the JSONL in a community batch import file is parsed, it's validated and errors are
+    recorded by line number and the error thrown, and this item represents that information,
+    which is returned to the uploading patron.
+    """
+
+    line_number: int
+    error_message: str
+
+
+@dataclass
+class CbiResult:
+    """
+    Represents the response item from community_batch_import().
+    """
+
+    batch: Batch | None = None
+    errors: list[CbiErrorItem] | None = None
+
+
+def format_pydantic_errors(
+    errors: list[ErrorDetails], line_idx: int
+) -> list[CbiErrorItem]:
+    """
+    Format a list of Pydantic errors into list[CbiErrorItem] for easier eventual HTML representation.
+    """
+    formatted_errors = []
+    for error in errors:
+        if loc := error.get("loc"):
+            loc_str = ", ".join(map(str, loc))
+        else:
+            loc_str = ""
+        msg = error["msg"]
+        error_type = error["type"]
+        error_message = f"{loc_str} field: {msg}. (Type: {error_type})"
+        formatted_errors.append(
+            CbiErrorItem(line_number=line_idx, error_message=error_message)
+        )
+    return formatted_errors
+
+
+def community_batch_import(raw_data: bytes) -> CbiResult:
+    """
+    This process raw byte data from a JSONL POST to the community batch import endpoint.
+
+    Each line in the JSONL file (i.e. import item) must pass the same validation as
+    required for a valid import going through load().
+
+    The return value has `batch` and `errors` attributes for the caller to use to
+    access the batch and any errors, respectively. See CbiResult for more.
+
+    The line numbers errors use 1-based counting because they are line numbers in a file.
+    """
+    user = accounts.get_current_user()
+    username = user.get_username()
+    batch_name = f"cb-{generate_hash(raw_data)}"
+    errors: list[CbiErrorItem] = []
+    dict_data = []
```

Comment:
I made this change, but I was hesitant to call them "records" or "recs" because they're not in the same format as what we elsewhere refer to import records, which is what gets returned by `edition_builder.get_dict()`. What about `raw_import_records` or something, so people are alerted they're _not_ "rec" items?

Reply:
Nice `raw_import_records` sounds perfect!